### PR TITLE
Refactor getting replication updates from database.

### PR DESCRIPTION
**Pull request by erikjohnston**
The aim here is to make it easier to reason about when streams are limited and when they're not, by moving the logic into the database functions themselves. This should mean we can kill of `db_query_to_update_function` function.

You might want to look at each commit individually.

This only does a subset of the streams, and simply merges the functionality from `db_query_to_update_function` into each streams get update DB function: https://github.com/matrix-org/synapse/blob/6c1f7c722f0baade9aecf41f600fcced670c4fcb/synapse/replication/tcp/streams/_base.py#L201-L218

c.f #7340

### CODE REVIEW

File: synapse/storage/data_stores/main/push_rule.py [hunk 1]
```diff
@@ -348,23 +348,33 @@ def bulk_get_push_rules_enabled(self, user_ids):
...
-            sql = (
-                "SELECT stream_id, event_stream_ordering, user_id, rule_id,"
-                " op, priority_class, priority, conditions, actions"
-                " FROM push_rules_stream"
-                " WHERE ? < stream_id AND stream_id <= ?"
-                " ORDER BY stream_id ASC LIMIT ?"
-            )
+            sql = """
+                SELECT stream_id, user_id
+                FROM push_rules_stream
```
Comment (richvdh) [inline synapse/storage/data_stores/main/push_rule.py:361]: why are we stuffing all those columns into the table if we don't care about them... :confused:

File: synapse/replication/tcp/streams/_base.py [hunk 2]
```diff
@@ -291,9 +291,7 @@ def __init__(self, hs):
         if hs.config.worker_app is None:
             # on the master, query the presence handler
             presence_handler = hs.get_presence_handler()
-            update_function = db_query_to_update_function(
-                presence_handler.get_all_presence_updates
-            )
+            update_function = presence_handler.get_all_presence_updates
```
Comment (richvdh) [inline synapse/replication/tcp/streams/_base.py:294]: maybe we should get rid of `presence_handler.get_all_presence_updates`, since it's now just a passthrough, and go straight to the storage layer, as many of the other streams do?

File: synapse/storage/data_stores/main/events_worker.py [hunk 3]
```diff
@@ -1077,9 +1077,14 @@ def get_ex_outlier_stream_rows_txn(txn):
             "get_ex_outlier_stream_rows", get_ex_outlier_stream_rows_txn
         )
 
-    def get_all_new_backfill_event_rows(self, last_id, current_id, limit):
+    async def get_all_new_backfill_event_rows(
+        self, instance_name: str, last_id: int, current_id: int, limit: int
+    ) -> Tuple[List[Tuple[int, list]], int, bool]:
+        """Get updates for backfill replication stream, including all new
+        backfilled events and events that have gone from being outliers to not.
+        """
```
Comment (richvdh) [inline synapse/storage/data_stores/main/events_worker.py:1085]: could you document the args and return format of the update functions? (Are the ids inclusive or exclusive?)

I know they are all much the same, but if I'm working on the storage layer, I don't want to have to go digging into how the replication layer works to know what a given function is supposed to do, and not having it written down explicitly is a good way for assumptions to be made and off-by-one errors to get in.

File: synapse/handlers/typing.py [hunk 4]
```diff
@@ -280,9 +280,9 @@ async def get_all_typing_updates(
             serial = self._room_serials[room_id]
             if last_id < serial <= current_id:
                 typing = self._room_typing[room_id]
-                rows.append((serial, room_id, list(typing)))
+                rows.append((serial, (room_id, list(typing))))
         rows.sort()
-        return rows[:limit]
+        return rows[:limit], current_id, False
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/typing.py:285]: Woops, for some reason i thought we weren't limiting for this

File: synapse/storage/data_stores/main/push_rule.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/storage/data_stores/main/push_rule.py:361]: Oh, hmm, it does look like they're not used anywhere else either. Though I don't really propose doing anything in this PR.

File: synapse/replication/tcp/streams/_base.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/replication/tcp/streams/_base.py:294]: Yeah, maybe. Though we really we should probably fix the `TODO` and correctly sync in memory updates as well, which means its sort of correct that we're going via the presence handler

File: synapse/handlers/typing.py [hunk 5]
```diff
@@ -282,7 +282,7 @@ async def get_all_typing_updates(
                 typing = self._room_typing[room_id]
                 rows.append((serial, (room_id, list(typing))))
         rows.sort()
-        return rows[:limit], current_id, False
+        return rows[:limit], current_id, len(rows) > limit
```
Comment (richvdh) [inline synapse/handlers/typing.py:285]: this is doing the wrong thing for the returned token when the limit is hit.

File: synapse/storage/data_stores/main/events_worker.py [hunk 6]
```diff
@@ -1082,6 +1082,21 @@ async def get_all_new_backfill_event_rows(
     ) -> Tuple[List[Tuple[int, list]], int, bool]:
         """Get updates for backfill replication stream, including all new
         backfilled events and events that have gone from being outliers to not.
+
+        Args:
+            instance_name: The writer we want to fetch updates from. Unused
+                here sincethere is only ever one writer.
```
Comment (richvdh) [inline synapse/storage/data_stores/main/events_worker.py:1088]:
```suggestion
                here since there is only ever one writer.
```

File: synapse/storage/data_stores/main/events_worker.py [hunk 7]
```diff
@@ -1082,6 +1082,21 @@ async def get_all_new_backfill_event_rows(
...
+            current_id: The token to fetch updates up to. Inclusive.
+            limit: The requested limit for the number of rows to return. The
+                function may return more or fewer rows.
+
+        Returns:
+            A tuple consisting of: the updates, the position of the rows
+            returned up to, and whether we returned fewer rows than exists
+            between the requested tokens due to the limit.
+
+            The updates are list of 2-tuples of stream ID and the row.
```
Comment (richvdh) [inline synapse/storage/data_stores/main/events_worker.py:1099]:
```suggestion
            The updates are a list of 2-tuples of stream ID and the row data
```

File: synapse/storage/data_stores/main/events_worker.py [hunk 8]
```diff
@@ -1082,6 +1082,21 @@ async def get_all_new_backfill_event_rows(
...
+            instance_name: The writer we want to fetch updates from. Unused
+                here sincethere is only ever one writer.
+            last_id: The token to fetch updates from. Exclusive.
+            current_id: The token to fetch updates up to. Inclusive.
+            limit: The requested limit for the number of rows to return. The
+                function may return more or fewer rows.
+
+        Returns:
+            A tuple consisting of: the updates, the position of the rows
+            returned up to, and whether we returned fewer rows than exists
```
Comment (richvdh) [inline synapse/storage/data_stores/main/events_worker.py:1096]: can you rephrase "the position of the rows returned up to"? it's somewhat unclear: inconsistent use of "position" and "token", inclusive or exclusive, etc.

"the last token included in the results", maybe?

File: synapse/storage/data_stores/main/events_worker.py [hunk 9]
```diff
@@ -1077,9 +1077,29 @@ def get_ex_outlier_stream_rows_txn(txn):
             "get_ex_outlier_stream_rows", get_ex_outlier_stream_rows_txn
         )
 
-    def get_all_new_backfill_event_rows(self, last_id, current_id, limit):
+    async def get_all_new_backfill_event_rows(
+        self, instance_name: str, last_id: int, current_id: int, limit: int
+    ) -> Tuple[List[Tuple[int, list]], int, bool]:
+        """Get updates for backfill replication stream, including all new
```
Comment (richvdh) [inline synapse/storage/data_stores/main/events_worker.py:1083]: this docstring is very helpful, but please can *all* the updated/new storage methods have one, not just this method?

File: synapse/replication/tcp/streams/_base.py (hunk 2 above)
Comment (richvdh) [reply synapse/replication/tcp/streams/_base.py:294]: fair

File: synapse/handlers/typing.py (hunk 5 above)
Comment (erikjohnston) [reply synapse/handlers/typing.py:285]: I am such a crank :facepalm:

File: synapse/storage/data_stores/main/events_worker.py (hunk 8 above)
Comment (erikjohnston) [reply synapse/storage/data_stores/main/events_worker.py]: Hopefully I've clarified it. Unfortunately, technically, the returned token doesn't *have* to be the last token included in the results (since `current_id` could be greater due to a gap)

File: synapse/handlers/typing.py (hunk 5 above)
Comment (erikjohnston) [reply synapse/handlers/typing.py]: MAYBE THIS TIME I'VE fIXeD IT?!!!??!1?
